File: mobile/lib/core/models/workout.dart
/// Modèles Workout — SOMA LOT 6.
///
/// ExerciseLibrary, WorkoutSet, ExerciseEntry, WorkoutSession.
library;

// ── Bibliothèque d'exercices ──────────────────────────────────────────────────

class ExerciseLibrary {
  final String id;
  final String name;
  final String? nameFr;
  final String category;
  final String? subcategory;
  final List<String> muscleGroups;
  final List<String> equipmentNeeded;
  final String difficultyLevel;
  final String? description;

  const ExerciseLibrary({
    required this.id,
    required this.name,
    this.nameFr,
    required this.category,
    this.subcategory,
    required this.muscleGroups,
    required this.equipmentNeeded,
    required this.difficultyLevel,
    this.description,
  });

  factory ExerciseLibrary.fromJson(Map<String, dynamic> json) =>
      ExerciseLibrary(
        id: json['id'] as String,
        name: json['name'] as String,
        nameFr: json['name_fr'] as String?,
        category: json['category'] as String? ?? 'other',
        subcategory: json['subcategory'] as String?,
        muscleGroups:
            (json['muscle_groups'] as List<dynamic>? ?? []).cast<String>(),
        equipmentNeeded:
            (json['equipment_needed'] as List<dynamic>? ?? []).cast<String>(),
        difficultyLevel:
            json['difficulty_level'] as String? ?? 'intermediate',
        description: json['description'] as String?,
      );

  String get displayName => nameFr ?? name;

  String get categoryLabel {
    switch (category) {
      case 'strength':
        return 'Force';
      case 'cardio':
        return 'Cardio';
      case 'hiit':
        return 'HIIT';
      case 'yoga':
        return 'Yoga';
      case 'mobility':
        return 'Mobilité';
      case 'core':
        return 'Gainage';
      default:
        return category;
    }
  }
}

// ── Série ─────────────────────────────────────────────────────────────────────

class WorkoutSet {
  final String id;
  final int setNumber;
  final String setType; // warmup|working|dropset|amrap|failure
  final int? repsTarget;
  final int? repsActual;
  final double? weightKg;
  final int? durationSeconds;
  final double? rpeSet;
  final bool isPr;
  final bool isDeleted;

  const WorkoutSet({
    required this.id,
    required this.setNumber,
    required this.setType,
    this.repsTarget,
    this.repsActual,
    this.weightKg,
    this.durationSeconds,
    this.rpeSet,
    this.isPr = false,
    this.isDeleted = false,
  });

  factory WorkoutSet.fromJson(Map<String, dynamic> json) => WorkoutSet(
        id: json['id'] as String,
        setNumber: json['set_number'] as int,
        setType: json['set_type'] as String? ?? 'working',
        repsTarget: json['reps_target'] as int?,
        repsActual: json['reps_actual'] as int?,
        weightKg: (json['weight_kg'] as num?)?.toDouble(),
        durationSeconds: json['duration_seconds'] as int?,
        rpeSet: (json['rpe_set'] as num?)?.toDouble(),
        isPr: json['is_pr'] as bool? ?? false,
        isDeleted: json['is_deleted'] as bool? ?? false,
      );

  String get display {
    final reps = repsActual ?? repsTarget;
    final weight = weightKg != null
        ? ' × ${weightKg!.toStringAsFixed(weightKg! == weightKg!.roundToDouble() ? 0 : 1)} kg'
        : '';
    if (reps != null) return '$reps reps$weight';
    if (durationSeconds != null) return '${durationSeconds}s$weight';
    return 'Set $setNumber';
  }
}

// ── Exercice dans une séance ──────────────────────────────────────────────────

class ExerciseEntry {
  final String id;
  final String exerciseId;
  final String exerciseName;
  final int exerciseOrder;
  final List<WorkoutSet> sets;
  final double tonnage;
  final int totalSets;
  final int totalReps;

  const ExerciseEntry({
    required this.id,
    required this.exerciseId,
    required this.exerciseName,
    required this.exerciseOrder,
    required this.sets,
    required this.tonnage,
    required this.totalSets,
    required this.totalReps,
  });

  factory ExerciseEntry.fromJson(Map<String, dynamic> json) {
    final setsRaw = json['sets'] as List<dynamic>? ?? [];
    return ExerciseEntry(
      id: json['id'] as String,
      exerciseId: json['exercise_id'] as String,
      exerciseName:
          json['exercise_name'] as String? ?? json['exercise_id'] as String,
      exerciseOrder: json['exercise_order'] as int? ?? 0,
      sets: setsRaw
          .map((s) => WorkoutSet.fromJson(s as Map<String, dynamic>))
          .where((s) => !s.isDeleted)
          .toList(),
      tonnage: (json['tonnage'] as num?)?.toDouble() ?? 0,
      totalSets: json['total_sets'] as int? ?? 0,
      totalReps: json['total_reps'] as int? ?? 0,
    );
  }
}

// ── Séance ────────────────────────────────────────────────────────────────────

class WorkoutSession {
  final String id;
  final String sessionType;
  final String location;
  final String status;
  final String startedAt;
  final String? endedAt;
  final int? durationMinutes;
  final int? rpeScore;
  final String? notes;
  final double? totalTonnageKg;
  final int? totalSets;
  final int? totalReps;
  final double? trainingLoad;
  final List<ExerciseEntry> exercises;

  const WorkoutSession({
    required this.id,
    required this.sessionType,
    required this.location,
    required this.status,
    required this.startedAt,
    this.endedAt,
    this.durationMinutes,
    this.rpeScore,
    this.notes,
    this.totalTonnageKg,
    this.totalSets,
    this.totalReps,
    this.trainingLoad,
    this.exercises = const [],
  });

  factory WorkoutSession.fromJson(Map<String, dynamic> json) {
    final exercisesRaw = json['exercises'] as List<dynamic>? ?? [];
    return WorkoutSession(
      id: json['id'] as String,
      sessionType: json['session_type'] as String,
      location: json['location'] as String? ?? 'gym',
      status: json['status'] as String? ?? 'planned',
      startedAt: json['started_at'] as String,
      endedAt: json['ended_at'] as String?,
      durationMinutes: json['duration_minutes'] as int?,
      rpeScore: json['rpe_score'] as int?,
      notes: json['notes'] as String?,
      totalTonnageKg: (json['total_tonnage_kg'] as num?)?.toDouble(),
      totalSets: json['total_sets'] as int?,
      totalReps: json['total_reps'] as int?,
      trainingLoad: (json['training_load'] as num?)?.toDouble(),
      exercises: exercisesRaw
          .map((e) => ExerciseEntry.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  String get typeLabel {
    switch (sessionType) {
      case 'strength':
        return 'Force';
      case 'cardio':
        return 'Cardio';
      case 'hiit':
        return 'HIIT';
      case 'yoga':
        return 'Yoga';
      case 'mobility':
        return 'Mobilité';
      case 'sport':
        return 'Sport';
      default:
        return 'Autre';
    }
  }

  String get statusLabel {
    switch (status) {
      case 'planned':
        return 'Planifiée';
      case 'in_progress':
        return 'En cours';
      case 'completed':
        return 'Terminée';
      case 'cancelled':
        return 'Annulée';
      default:
        return status;
    }
  }

  String get locationLabel {
    switch (location) {
      case 'gym':
        return 'Salle';
      case 'home':
        return 'Maison';
      case 'outdoor':
        return 'Extérieur';
      default:
        return 'Autre';
    }
  }

  bool get isCompleted => status == 'completed';
  bool get isInProgress => status == 'in_progress';
}

File: backend/app/models/workout.py
import uuid
from typing import Optional, List
from datetime import datetime
from sqlalchemy import ForeignKey, String, Boolean, DateTime, func, Float, Integer, Text, ARRAY, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID, JSONB

from app.db.base import Base, UUIDMixin

# Valeurs valides pour WorkoutSession.status
SESSION_STATUS_VALUES = ("planned", "in_progress", "completed", "skipped", "cancelled")


class ExerciseLibrary(Base, UUIDMixin):
    """Bibliothèque d'exercices — Module E."""
    __tablename__ = "exercise_library"

    name: Mapped[str] = mapped_column(String(200), nullable=False)
    name_fr: Mapped[Optional[str]] = mapped_column(String(200))
    slug: Mapped[Optional[str]] = mapped_column(String(100), unique=True)
    category: Mapped[Optional[str]] = mapped_column(String(50))  # strength, cardio, mobility, balance, hiit
    subcategory: Mapped[Optional[str]] = mapped_column(String(50))
    primary_muscles: Mapped[Optional[List[str]]] = mapped_column(ARRAY(Text))
    secondary_muscles: Mapped[Optional[List[str]]] = mapped_column(ARRAY(Text))
    difficulty_level: Mapped[Optional[str]] = mapped_column(String(20))  # beginner, intermediate, advanced
    equipment_required: Mapped[Optional[List[str]]] = mapped_column(ARRAY(Text))
    execution_location: Mapped[Optional[str]] = mapped_column(String(20))  # gym, home, outdoor, any
    description: Mapped[Optional[str]] = mapped_column(Text)
    instructions: Mapped[Optional[List[str]]] = mapped_column(ARRAY(Text))
    breathing_cues: Mapped[Optional[str]] = mapped_column(Text)
    common_errors: Mapped[Optional[List[str]]] = mapped_column(ARRAY(Text))
    image_urls: Mapped[Optional[List[str]]] = mapped_column(ARRAY(Text))
    video_url: Mapped[Optional[str]] = mapped_column(String(500))
    easier_variant_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True))
    harder_variant_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True))
    key_joint_angles: Mapped[Optional[dict]] = mapped_column(JSONB)
    cv_supported: Mapped[bool] = mapped_column(Boolean, default=False)
    rep_detection_model: Mapped[Optional[str]] = mapped_column(String(50))
    contraindications: Mapped[Optional[List[str]]] = mapped_column(ARRAY(Text))
    met_value: Mapped[Optional[float]] = mapped_column(Float)
    format_type: Mapped[Optional[str]] = mapped_column(String(20))  # reps | duration | distance
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())


class WorkoutSession(Base, UUIDMixin):
    """Session d'entraînement — Module G."""
    __tablename__ = "workout_sessions"

    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    started_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False, index=True)
    ended_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    duration_minutes: Mapped[Optional[int]] = mapped_column(Integer)
    session_type: Mapped[Optional[str]] = mapped_column(String(50))
    # strength | cardio | hiit | mobility | walk | elliptical | yoga | swimming | bodyweight | mat

    # --- AJOUT LOT 1 : statut de la séance et soft delete ---
    status: Mapped[str] = mapped_column(String(20), nullable=False, default="planned", server_default="planned")
    # planned | in_progress | completed | skipped | cancelled
    is_deleted: Mapped[bool] = mapped_column(Boolean, default=False, server_default="false")

    location: Mapped[Optional[str]] = mapped_column(String(50))  # gym, home, outdoor

    # Volume (recalculé à chaque ajout de set)
    total_tonnage_kg: Mapped[Optional[float]] = mapped_column(Float)
    total_sets: Mapped[Optional[int]] = mapped_column(Integer)
    total_reps: Mapped[Optional[int]] = mapped_column(Integer)

    # Cardio
    distance_km: Mapped[Optional[float]] = mapped_column(Float)
    avg_heart_rate_bpm: Mapped[Optional[float]] = mapped_column(Float)
    max_heart_rate_bpm: Mapped[Optional[float]] = mapped_column(Float)
    calories_burned_kcal: Mapped[Optional[float]] = mapped_column(Float)

    # Charge d'entraînement
    internal_load_score: Mapped[Optional[float]] = mapped_column(Float)   # durée × RPE
    rpe_score: Mapped[Optional[float]] = mapped_column(Float)              # 1-10

    # Ressenti
    energy_before: Mapped[Optional[int]] = mapped_column(Integer)          # 1-10
    energy_after: Mapped[Optional[int]] = mapped_column(Integer)
    perceived_difficulty: Mapped[Optional[int]] = mapped_column(Integer)
    technical_score: Mapped[Optional[float]] = mapped_column(Float)        # score qualité technique (CV)

    notes: Mapped[Optional[str]] = mapped_column(Text)
    is_completed: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    # Relations
    exercises: Mapped[List["WorkoutExercise"]] = relationship(
        "WorkoutExercise",
        back_populates="session",
        order_by="WorkoutExercise.exercise_order",
        primaryjoin="and_(WorkoutExercise.session_id == WorkoutSession.id, WorkoutExercise.is_deleted == False)",
    )


class WorkoutExercise(Base, UUIDMixin):
    """Exercice dans une session d'entraînement."""
    __tablename__ = "workout_exercises"

    session_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("workout_sessions.id", ondelete="CASCADE"), nullable=False, index=True)
    exercise_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True))
    exercise_order: Mapped[int] = mapped_column(Integer, default=1)
    notes: Mapped[Optional[str]] = mapped_column(Text)
    biomechanics_score: Mapped[Optional[float]] = mapped_column(Float)
    is_deleted: Mapped[bool] = mapped_column(Boolean, default=False, server_default="false")
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    # Relations
    session: Mapped["WorkoutSession"] = relationship("WorkoutSession", back_populates="exercises", foreign_keys=[session_id])
    sets: Mapped[List["WorkoutSet"]] = relationship(
        "WorkoutSet",
        back_populates="exercise",
        order_by="WorkoutSet.set_number",
        primaryjoin="and_(WorkoutSet.workout_exercise_id == WorkoutExercise.id, WorkoutSet.is_deleted == False)",
    )


class WorkoutSet(Base, UUIDMixin):
    """Série dans un exercice."""
    __tablename__ = "workout_sets"

    workout_exercise_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("workout_exercises.id", ondelete="CASCADE"), nullable=False, index=True)
    set_number: Mapped[int] = mapped_column(Integer, nullable=False)
    reps_target: Mapped[Optional[int]] = mapped_column(Integer)
    reps_actual: Mapped[Optional[int]] = mapped_column(Integer)
    weight_kg: Mapped[Optional[float]] = mapped_column(Float)
    duration_seconds: Mapped[Optional[int]] = mapped_column(Integer)   # Pour iso / cardio
    rest_seconds: Mapped[Optional[int]] = mapped_column(Integer)
    tempo: Mapped[Optional[str]] = mapped_column(String(20))           # '3-1-2'
    time_under_tension_s: Mapped[Optional[float]] = mapped_column(Float)
    range_of_motion_pct: Mapped[Optional[float]] = mapped_column(Float)
    rpe_set: Mapped[Optional[float]] = mapped_column(Float)
    is_warmup: Mapped[bool] = mapped_column(Boolean, default=False)
    is_pr: Mapped[bool] = mapped_column(Boolean, default=False)
    is_deleted: Mapped[bool] = mapped_column(Boolean, default=False, server_default="false")
    data_source: Mapped[str] = mapped_column(String(20), default="manual")  # manual | camera | estimated
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    # Relations
    exercise: Mapped["WorkoutExercise"] = relationship("WorkoutExercise", back_populates="sets", foreign_keys=[workout_exercise_id])
